db/migrate/20180910155506_add_uniqueness_of_order_id_to_spree_shipments.rb

```diff
@@ -1,6 +1,23 @@
+# This migration is an OFN specific migration that enforces an order to have a single shipment at all times
 class AddUniquenessOfOrderIdToSpreeShipments < ActiveRecord::Migration
   def change
+    destroy_all_but_latest_shipments
+
     remove_index :spree_shipments, :order_id
     add_index :spree_shipments, :order_id, unique: true
   end
+
+  private
+
+  # Destroy all but the latest shipment in each order
+  def destroy_all_but_latest_shipments
+    latest_shipments = Spree::Shipment.
+      select("order_id, MAX(updated_at) updated_at").
+      group(:order_id).
+      having("count(*) > 1")
+
+    Spree::Shipment.
+      joins("INNER JOIN (#{latest_shipments.to_sql}) latest_shipments ON spree_shipments.order_id=latest_shipments.order_id AND spree_shipments.updated_at != latest_shipments.updated_at").
+      destroy_all
```

Comment:
It is unlikely for two shipments for the same order to have the same `updated_at`, but it is still possible.

1. Can we also check for the maximum ID to handle this case?
2. Should we consider the shipment state, e.g. prefer a shipment that is "shipped" over one that is "ready" or "pending"? (State machine [here](https://github.com/openfoodfoundation/spree/blob/step-6a/core/app/models/spree/shipment.rb#L66))

And sorry if this was already discussed before: Do we have a system for validating a certain v1 instance will not lose any data when migrating? For example, for this PR, if there was an order for which I had to ship twice, I might lose tracking data for the first shipment. Are we okay with this?

Reply:
I have checked live data before doing this and we are talking about around 16 orders in each of main environments (AU, FR and UK). Each environment has 2/3 orders like this in 2019...
All the duplicate shipments of an order have the same workflow state. All the shipments in the orders with duplicate shipments have different updated_at, for example:
```
 29217 |          | H86046673735 |      |                            |   398707 |                 46 |     269274 | 2018-11-04 01:58:55.405496 | 2018-11-04 01:58:55.405496 | pending
 29218 |          | H46481084155 |      |                            |   398707 |                 46 |     269276 | 2018-11-04 01:58:55.690717 | 2018-11-04 01:58:55.690717 | pending
```
I have checked some addresses of these duplicates and they are also the same... the shipping_methods are also the same for the duplicates I checked.
So, I think your point is very valid but not applicable here. I dont think we need to spend more time with this. what do you think?

regarding your last question:
no, we dont have any automatic validation in place, we just make sure that in every PR we keep the data consistent.
I dont think we will have any trouble with this particular PR.
do you have some other part of the model in mind where we should think about this? It's an important topic...

Reply:
It'd be great to get an idea of how long this takes with the snapshot we are currently working with. Unless it takes too long I wouldn't care much about `updated_at` or `id` or any other improvement.

We also need to be aware that we're instantiating the shipments to destroy them afterward (executing callbacks) and that takes time. Again, if it is quick enough I wouldn't care much.

Reply:
we are talking about changing 20/40 records in a table with 20k/30k records. I dont think it will be a problem.

Reply:
I think it's bad practice to assume that the latest shipment is the one that we should keep. And actually, I'm uncomfortable with the idea of losing the older shipment data too.

Relating this with @mkllnk's recommendation for a CSV export before. What about going ahead with the code to keep the latest shipment record, but exporting to a CSV the shipment data of the ones we destroy? The instances can then decide if they want to contact the affected enterprises - @luisramos0 you said there are few so it would be easy enough for them.

Reply:
ok, I think your concern is very important, I moved the generic question to the data migration epic [here](https://github.com/openfoodfoundation/openfoodnetwork/issues/3285#issuecomment-476709344).

it's not just bad practice, this would be totally unacceptable in another contexts (including other areas of the OFN model). but the further I look into the live data the more I think we dont need to worry here.

I'll consider our 4 main envinroments here.
I created a few queries.

#### duplicate shipments (multiple shipments in the same order)
```
select * from spree_shipments
INNER JOIN (select order_id from spree_shipments group by order_id having count(*) > 1) orders_with_duplicates ON spree_shipments.order_id = orders_with_duplicates.order_id;
```
This shows a total of 114 duplicates (~50 orders in total)
- fr live 39 rows
- uk prod 33 rows
- katuma prod 2 rows
- aus prod 40 rows

from these, I checked what were the ones where shipping_method or state was different between duplicates:

#### duplicate spree shipments with different states or shipping method
```
select * from spree_shipments
INNER JOIN (select order_id from spree_shipments group by order_id having count(*) > 1) orders_with_duplicates ON spree_shipments.order_id = orders_with_duplicates.order_id
where exists (select id from spree_shipments sp where sp.order_id = spree_shipments.order_id and sp.id != spree_shipments.id and (sp.state != spree_shipments.state OR sp.shipping_method_id != spree_shipments.shipping_method_id));

```
- FR prod - no entries
- in uk prod - 3 orders where the shipping method changed (one order in 2018, from Sept)
- katuma prod - no entries
- AUS prod - 5 orders where the shipping method changed (one order in 2018, from Nov)

and then,
#### duplicated shipments with address diferences
```
select * from spree_shipments
INNER JOIN (select order_id from spree_shipments group by order_id having count(*) > 1) orders_with_duplicates ON spree_shipments.order_id = orders_with_duplicates.order_id
where exists (select sp.id from spree_shipments sp
  inner join spree_addresses a1 on (a1.id = sp.address_id)
  inner join spree_addresses a2 on (a2.id = spree_shipments.address_id)
  where sp.order_id = spree_shipments.order_id and sp.id != spree_shipments.id and sp.address_id != spree_shipments.address_id and
    (
        a1.firstname != a2.firstname OR a1.lastname != a2.lastname OR a1.address1 != a2.address1 OR a1.address2 != a2.address2 OR a1.city != a2.city
        OR a1.zipcode != a2.zipcode OR a1.phone != a2.phone OR a1.state_id != a2.state_id OR a1.country_id != a2.country_id
        OR a1.company != a2.company OR a1.latitude != a2.latitude OR a1.longitude != a2.longitude OR a1.state_name != a2.state_name
        OR a1.alternative_phone != a2.alternative_phone
    )
  );
```
and this is empty in all environments.

So, we are really talking about 8 orders (2 from 2018) where the shipping method was updated.

Reply:
I think you suggestion is acceptable, it just takes time...
I'll add a new commit with a backup process shortly.

Reply:
ok, done [here](3742c70910897c2069ae5afc35e1778c5ae879ee).
Let me know if this was what you were thinking about.
Thanks!

Reply:
@luisramos0 I didn't realize that the "H86046673735" and "H46481084155" in your examples above are Spree-generated shipment numbers - I thought they were user-input tracking numbers. :joy: 

Nevertheless, I still think we should extract these data because of the fact that the users could have deleted the shipments but they did not.

Reply:
@luisramos0 I wondered why the users did not remove the previous shipments, and realized it could be because these shipments were associated with shipping fee adjustments: `source_type: "Spree::Shipment"` ...

Do the older shipments have associated adjustments? If yes, maybe it will work to just reassign the adjustments to the latest shipments?

Reply:
I checked the db and there are adjustments for all shipments, most of the amounts are zero.
in FR and UK, the duplicate shipments have all the same adjustment. in AUS, only in 3 orders we have adjustments with different amounts.

I think I can add to the migration:
- backup all the adjustments of duplicate shipments
- delete all the adjustments of the shipments that are being deleted

is this acceptable for you? let me know.

Reply:
H86046673735 is the shipment number, tracking is another field in the shipment.

Reply:
I have added one more commit to backup adjustments with the shipments and delete the adjustments before deleting the duplicate shipments.

Reply:
@luisramos0 My preference is to keep the adjustment but make it like a custom unassociated adjustment (what you get when using "New Adjustment" button). From a quick test adding such an adjustment, seeing that `source`, `originator`, and `mandatory` are blank. What do you think about this approach?

If doing this way, I think we don't need to export affected adjustments.

The intention is to preserve as much of the users' current data  as we can.

Reply:
I am not sure I understand , do you mean to update the adjustments of the deleted shipments and some fields to blank? If that's what you suggest, I dont like the idea of manipulating the data, things can break. I prefer to delete them.
But if you prefer the update, please be a bit more specific about your suggestion and I'll implement it, ok?

Reply:
My idea was to do it like this, @luisramos0:

```ruby
adjustments_scope.update_all(source_id: nil, source_type: nil, originator_id: nil,
                             originator_type: nil, mandatory: nil)
```

That just removes their association with the shipment and shipping method, instead of deleting the adjustments.

Reply:
ok, I replaced the last commit with that.
I tested with uk live data and it works.
this is the SQL query:
```
UPDATE "spree_adjustments"
SET "source_id" = NULL, "source_type" = NULL, "originator_id" = NULL, "originator_type" = NULL, "mandatory" = NULL
WHERE "spree_adjustments"."id" IN
(SELECT "spree_adjustments"."id"
FROM "spree_adjustments"
INNER JOIN (SELECT "spree_shipments".*
          FROM "spree_shipments"
          INNER JOIN (
                 SELECT order_id, MAX(updated_at) updated_at
                  FROM "spree_shipments"
                  GROUP BY order_id HAVING count(*) > 1) latest_shipments
             ON spree_shipments.order_id = latest_shipments.order_id
                 AND spree_shipments.updated_at != latest_shipments.updated_at) shipments_to_delete
  ON shipments_to_delete.id = spree_adjustments.source_id
     and spree_adjustments.source_type = 'Spree::Shipment')
```